### ISSUE TITLE
Documentation should use "write_all" instead of "write"

### ISSUE THREAD

**Issue by evanw** [alexcrichton/flate2-rs#151]
I tried using this library like this:

```rust
let mut encoder = flate2::write::DeflateEncoder::new(vec![], flate2::Compression::default());
encoder.write(bytes).unwrap();
let deflated = encoder.finish().unwrap();
```

That turned out to silently generate corrupted data sometimes. This way appears to work:

```rust
let mut encoder = flate2::read::DeflateEncoder::new(bytes, flate2::Compression::default());
let mut deflated = vec![];
encoder.read_to_end(&mut deflated).unwrap();
```

Can you explain the difference? Was I wrong to do it the first way? It seemed to be the approach suggested by the readme.

**Comment by alexcrichton** [comment]
Oh dear, this may indeed be a bug! Can you provide the input which causes the weirdness?

**Comment by evanw** [comment]
Sure. First, download [data.txt](https://github.com/alexcrichton/flate2-rs/files/1858103/data.txt). This contains the base64-encoded data that caused the problem. Then:

```
$ base64 -D data.txt > data.bin

$ shasum data.bin
fde480b272d08d32e6fad01c0a9253c4b4bc8f83  data.bin

$ mkdir src

$ cat > Cargo.toml
[package]
name = "test"
version = "0.0.1"

[dependencies]
flate2 = "1.0"

$ cat > src/main.rs
extern crate flate2;

use std::io::*;

fn deflate_bytes_write(bytes: &[u8]) -> Vec<u8> {
  let mut encoder = flate2::write::DeflateEncoder::new(vec![], flate2::Compression::default());
  encoder.write(bytes).unwrap();
  return encoder.finish().unwrap();
}

fn deflate_bytes_read(bytes: &[u8]) -> Vec<u8> {
  let mut encoder = flate2::read::DeflateEncoder::new(bytes, flate2::Compression::default());
  let mut deflated = vec![];
  encoder.read_to_end(&mut deflated).unwrap();
  return deflated;
}

fn inflate_bytes_read(bytes: &[u8]) -> Vec<u8> {
  let mut decoder = flate2::read::DeflateDecoder::new(bytes);
  let mut output = vec![];
  decoder.read_to_end(&mut output).unwrap();
  return output;
}

fn main() {
  let mut input = BufReader::new(stdin());
  let mut bytes = vec![];
  input.read_to_end(&mut bytes).unwrap();
  println!("input is {} bytes", bytes.len());

  let deflated = deflate_bytes_read(&bytes);
  println!("deflate_bytes_read gives {} bytes", deflated.len());
  assert_eq!(bytes.len(), inflate_bytes_read(&deflated).len());

  let deflated = deflate_bytes_write(&bytes);
  println!("deflate_bytes_write gives {} bytes", deflated.len());
  assert_eq!(bytes.len(), inflate_bytes_read(&deflated).len());
}

$ cargo run < data.bin
   Compiling cc v1.0.9
   Compiling libc v0.2.40
   Compiling miniz-sys v0.1.10
   Compiling flate2 v1.0.1
   Compiling test v0.0.1 (file:///Users/evan/Desktop/flate2)
    Finished dev [unoptimized + debuginfo] target(s) in 5.28 secs
     Running `target/debug/test`
input is 188718 bytes
deflate_bytes_read gives 53325 bytes
deflate_bytes_write gives 50088 bytes
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `188718`,
 right: `181477`', src/main.rs:37:2
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by alexcrichton** [comment]
Ah thanks for the data! I think in that script the `write` needs to be changed to `write_all` though to work correctly. After that though does this line up for you?

**Comment by oyvindln** [comment]
The [`write`](https://doc.rust-lang.org/std/io/trait.Write.html#tymethod.write) method on the `Writer` trait isn't guaranteed to write all the data input to it, the return value on success returns the actual number of bytes written. For that you need to call `write_all`. Maybe the method should have been named differently to indicate this, personally I find it a bit misleading, though that is an issue with the rust standard library rather than `flate2`. It would probably be a good idea for the `flate2` examples to use `write_all` to avoid any confusion.

So TLDR, I believe the behaviour is correct, but the docs could use a fix.

 `read_to_end` does, as the name suggests read to the end, so if it succeeds, it does pull everything from the reader as expected.

**Comment by evanw** [comment]
Oh wow, that is an unfortunate name. I can confirm that it works correctly for me now. Thanks for the help! Changing the documentation for this project to use `write_all` instead of `write` sounds like the right fix here. It looks like both the comments in the code and the readme file would need to be changed.
